src/main/cpp/native_glue.cc

```diff
@@ -188,41 +122,34 @@ namespace litecore {
             }
         }
 
-        jstringSlice::jstringSlice(JNIEnv *env, jstring js)
-                : _env(nullptr) {
+        jstringSlice::jstringSlice(JNIEnv *env, jstring js) {
             assert(env != nullptr);
             if (js != nullptr) {
-                jboolean isCopy;
-                _jstr = js;
-                _env = env;
-
-                char *cstr = (char *)env->GetStringUTFChars(js, &isCopy);
-                assert(isCopy);
-                ModifiedUTF8ToUTF8(cstr);
-                if (!cstr)
-                    return; // Would it be better to throw an exception?
+
+                // !!! COPYING THE STRING 2x?
+                auto str = JstringToUTF8(env, js);
+
+                auto buf = (char *) malloc(str.length() + 1);
+                if (buf == nullptr) {
+                    jclass c = env->FindClass("java/lang/OutOfMemoryError");
+                    env->ThrowNew(c, "Out of memory creating String slice");
+                }
+                char *cstr = strcpy(buf, str.c_str());;
+
                 _slice = slice(cstr);
             }
         }
 
         jstringSlice::~jstringSlice() {
-            if (_env)
-                _env->ReleaseStringUTFChars(_jstr, (const char *) _slice.buf);
-            else if (_slice.buf)
-                free((void *) _slice.buf);        // detached
+            free((void *) _slice.buf);
         }
 
+        // !!! DOES THIS STILL WORK?  IS IT STILL NECESSARY?
         void jstringSlice::copyAndReleaseRef() {
-            if (_env) {
-                auto cstr = (const char *) _slice.buf;
-                _slice = _slice.copy();
-                _env->ReleaseStringUTFChars(_jstr, cstr);
-                _env->DeleteLocalRef(_jstr);
-                _env = nullptr;
-            }
+            _slice = _slice.copy();
```

Comment:
I think we need to free the old _slice, right? Are we using this anywhere?

Reply:
I don't really know what this method is supposed to do.  I don't think it freed the old `_slice` before I changed it.  I suspect that it is an attempt to free `_jstr` so that the GC can collect it.  Since we don't hold a reference to the jstring anymore, I'm not sure that this is even useful.  Definitely could use your advice here.